sql/wsrep_encryption.cc

```diff
@@ -0,0 +1,136 @@
+ * Serialize key_version and key into buffer
+ * 
+ * @param input Container for serialized key
+ * @param key Pointer to encryption key
+ * @param size Size of key
+ * @param version Key version
+ */
+static void wsrep_key_serialize(std::vector<unsigned char>& input, const void* key,
+                                size_t size, unsigned int version)
+{
+  input.resize(key_version_preamble_size + size);
+  memcpy(input.data(), (void *)&version, sizeof(version));
+  memcpy(input.data() + key_version_preamble_size, key, size);
+}
+
+/**
+ * De-serialize buffer into key_version, key_size & key_ptr
+ * 
+ * @param input Pointer to serialized key
+ * @param size Size of serialized key
+ * @param key_ptr Pointer to begining of encryption key
+ * @param size Size of encryption key
+ * @param version Key version
+ * 
+ * @return 0 when key is sucessfuly de-serialized. 1 otherwise.
+ */
+static int wsrep_key_deserialize(const void* input, const size_t& input_size,
+                                 const void*& key_ptr, size_t& size,
+                                  unsigned int& version)
+{
+  bool ret= 1;
+  if(input_size >= key_version_preamble_size)
+  {
+    memcpy(&version, input, sizeof(version));
+    key_ptr= static_cast<const char*>(input) + key_version_preamble_size;
+    size= input_size - key_version_preamble_size;
+    ret= 0;
+  }
+  return ret;
+}
+
+int wsrep_set_encryption_key(const void* key, size_t size, unsigned int version)
+{
+  std::vector<unsigned char> input;
+  wsrep_key_serialize(input, key, size, version);
+  return Wsrep_server_state::instance().set_encryption_key(input);
+}
+
+
+int Wsrep_encryption_service::do_crypt(void**                ctx,
+                                       wsrep::const_buffer&  key,
+                                       const char            (*iv)[32],
+                                       wsrep::const_buffer&  input,
+                                       void*                 output,
+                                       bool                  encrypt,
+                                       bool                  last)
+{
+
+  const void* deserialized_key_ptr;
+  size_t deserialized_key_size;
+  unsigned int key_version;
+  if(wsrep_key_deserialize(key.data(), key.size(), deserialized_key_ptr,
+                           deserialized_key_size, key_version))
+  {
+     throw wsrep::runtime_error("Failed wsrep_key_deserialize()");
+  }
+  if (*ctx == NULL)
+  {
+    int mode= encrypt ? ENCRYPTION_FLAG_ENCRYPT : ENCRYPTION_FLAG_DECRYPT;
+    *ctx = ::malloc(encryption_ctx_size(ENCRYPTION_KEY_SYSTEM_DATA,
+                                        key_version));
+    if (*ctx == NULL)
+    {
+      throw wsrep::runtime_error("Memory not allocated in do_crypt()");
+    }
+    if(encryption_ctx_init(*ctx,
+                           (const unsigned char*)deserialized_key_ptr,
+                           deserialized_key_size,
+                           (unsigned char*)iv, MY_AES_BLOCK_SIZE,
```

Comment:
Cant we use unsigned char also on function parameters to avoid this C-style cast also ?